crates/protocol/src/execute_protocol.rs

```diff
@@ -85,64 +89,118 @@ pub async fn execute_protocol_generic<Res: synedrion::ProtocolResult>(
 
     loop {
         let mut accum = session.make_accumulator();
+        let current_round = session.current_round();
+        let session_arc = Arc::new(session);
+
+        // Send outgoing messages
+        let destinations = session_arc.message_destinations();
+        if !destinations.is_empty() {
+            // Channel for receiving message artifacts
+            let (artifact_tx, mut artifact_rx) = mpsc::channel(destinations.len());
+            for destination in destinations.iter() {
+                let session_arc = session_arc.clone();
+                let tx = tx.clone();
+                let my_id = my_id.clone();
+                let artifact_tx = artifact_tx.clone();
+                let destination = destination.clone();
+                tokio::spawn(async move {
+                    let result = match session_arc.make_message(&mut OsRng, &destination) {
+                        Ok((message, artifact)) => {
+                            match tx.send(ProtocolMessage::new(&my_id, &destination, message)) {
+                                Ok(_) => Ok(artifact),
+                                Err(err) => {
+                                    let err: GenericProtocolError<Res> = err.into();
+                                    Err(err)
+                                },
+                            }
+                        },
+                        Err(err) => Err(err.into()),
+                    };
+                    if artifact_tx.send(result).await.is_err() {
+                        tracing::error!("Protocol finished before outgoing message artifact sent");
+                    }
```

Comment:
```suggestion
                tokio::spawn(async move {
                    let result = session_arc
                        .make_message(&mut OsRng, &destination)
                        .map(|(message, artifact)| {
                            tx.send(ProtocolMessage::new(&my_id, &destination, message))
                                .map(|_| artifact)
                                .map_err(|err| {
                                    let err: GenericProtocolError<Res> = err.into();
                                    err
                                })
                        })
                        .map_err(|err| {
                            let err: GenericProtocolError<Res> = err.into();
                            err
                        });

                    if artifact_tx.send(result).await.is_err() {
                        tracing::error!("Protocol finished before outgoing message artifact sent");
                    }
                });
```

Got this to work, but not sure how much better it is tbh. Also had to made a small change below

Reply:
thank you!